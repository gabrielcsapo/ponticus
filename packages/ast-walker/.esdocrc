{
   "title": "typhonjs-ast-walker",
   "source": "src",
   "destination": "docs",

   "manual": { "changelog": ["./CHANGELOG.md"] }
}