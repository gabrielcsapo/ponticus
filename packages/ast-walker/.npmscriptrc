/**
 * You can provide comments in `.npmscriptrc`
 */
{
   "build":
   {
      "babel": { "source": "src", "destination": "dist" }
   },

   "publish":
   {
      "prepublish": { "scripts": ["npm run eslint", "npm run build", "npm run test"] }
   },

   "test":
   {
      // Provides a `coverage` handling command that is appended when running on Travis CI.
      "travis":
      {
         "istanbul": { "command": "cover", "options": ["--report lcovonly"] },
         "report": "./node_modules/.bin/codecov"
      },

      "istanbul": { "command": "cover" },
      "mocha": { "source": "./test/src", "options": [ "-u tdd", "--compilers js:babel-register", "-t 120000 --recursive" ] }
   }
}
