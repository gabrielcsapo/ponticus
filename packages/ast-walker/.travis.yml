language: node_js
node_js:
  - "5.8.0"
script:
  - npm run eslint
  - npm run build
  - npm run test-coverage