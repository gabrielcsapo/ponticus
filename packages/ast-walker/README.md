# @ponticus/ast-walker

Provides a simple Javascript AST traversal utility that traverses all nodes / children regardless of type.

A single method `traverse` takes an AST object or array of nodes and a callback object which may contain two methods `enterNode` and `exitNode` which are invoked with the current node and the parent node respectively when entering and exiting a given node during traversal.

`enterNode` may return a array of strings which provide a set of children keys to ignore or `null` to skip traversing children keys entirely.

To install `@ponticus/ast-walker` provide this entry in `package.json`:

```json
{
  ...
  "dependencies": {
    "@ponticus/ast-walker": "^0.2.0"
  }
}
```  

A simple example follows:

```js
import walker from '@ponticus/ast-walker';

const ast = .... // An AST tree.

walker.traverse(ast, {
   enterNode: (node, parent) => {
      console.log(`walker - enterNode: ${node.type}`);

      // Optional return statement to ignore specific children keys.
      // return node.type === 'ClassBody' ? ['body'] : undefined;

      // Optional return statement to ignore specific children keys or skip traversal entirely.
      // return node.type === 'ClassBody' ? null : undefined;
   },
   
   exitNode: (node, parent) => {
      console.log(`walker - exitNode: ${node.type}`);
   }
});
 
```

Support for [@ponticus/plugin-manager](https://www.npmjs.com/package/@ponticus/plugin-manager) is also available and
when loading `@ponticus/ast-walker` as a plugin it will automatically register event bindings on the plugin eventbus
with these event categories:

`ast:walker:traverse` - invokes `traverse` of the default walker.

Example `@ponticus/plugin-manager` usage:

```js
import PluginManager    from '@ponticus/plugin-manager';
import eventbus         from '@ponticus/backbone-esnext-eventbus';

const pluginManager = new PluginManager({ eventbus });

pluginManager.add({ name: @ponticus/ast-walker });

// This will automatically wire up @ponticus/ast-walker to the eventbus. 

eventbus.trigger('ast:walker:traverse', ast, {
   enterNode: (node, parent) => {
      console.log(`walker - enterNode: ${node.type}`);
   },
   exitNode: (node, parent) => {
      console.log(`walker - exitNode: ${node.type}`);
   }
});
```