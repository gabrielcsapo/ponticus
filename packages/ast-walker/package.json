{
  "name": "@ponticus/ast-walker",
  "version": "0.2.1",
  "homepage": "https://github.com/ponticus/ast-walker/",
  "description": "Provides a simple Javascript AST traversal utility that traverses all nodes / children regardless of type.",
  "main": "./dist",
  "repository": {
    "type": "git",
    "url": "https://github.com/typhonjs-node-ast/@ponticus/ast-walker.git"
  },
  "devDependencies": {
    "@swc/cli": "^0.1.57",
    "@swc/core": "^1.2.207",
    "@swc/jest": "^0.2.21",
    "eslint": "^8.18.0",
    "eslint-config-prettier": "^8.5.0",
    "vitest": "^0.16.0"
  },
  "scripts": {
    "build": "swc src -d dist",
    "lint": "eslint .",
    "prepublish": "npm run build",
    "test": "vitest"
  },
  "keywords": [
    "typhonjs",
    "ast",
    "traversal"
  ],
  "volta": {
    "node": "16.15.0"
  }
}
