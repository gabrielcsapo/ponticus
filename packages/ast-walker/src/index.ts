import { type PluginEvent } from "@ponticus/plugin-manager";

import ASTWalker from "@ponticus/ast-walker/ASTWalker";

export { ASTWalker };

/**
 * Default walker instance.
 */
const walker = new ASTWalker();

/**
 * Provides a default ASTWalker instance.
 */
export default walker;

/**
 * Wires up walker on the plugin eventbus.
 *
 * @param ev - The plugin event.
 *
 */
export function onPluginLoad(ev: PluginEvent) {
  let eventPrepend = "typhonjs";

  const options = ev.pluginOptions;

  // Apply any plugin options.
  if (typeof options === "object") {
    // If `eventPrepend` is defined then it is prepended before all event bindings.
    if (typeof options.eventPrepend === "string") {
      eventPrepend = `${options.eventPrepend}:`;
    }
  }

  ev.eventbus?.on(`${eventPrepend}:ast:walker:traverse`, walker.traverse);
}
