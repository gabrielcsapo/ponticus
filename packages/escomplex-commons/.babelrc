{
  "presets": ["env", "stage-2"],

  "plugins":
  [
    "add-module-exports",
    ["module-resolver", {
      "root": ["."],
      "alias":
      {
      }
    }],
    "transform-runtime"
  ],

  "env":
  {
    "escomplex-dev":
    {
      "plugins":
      [
        "add-module-exports",
        ["module-resolver", {
          "root": ["."],
          "cwd": "babelrc",
          "alias":
          {
            "typhonjs-escomplex-test-data": "../typhonjs-escomplex-test-data"
          }
        }],
        "transform-runtime"
      ]
    }
  }
}
