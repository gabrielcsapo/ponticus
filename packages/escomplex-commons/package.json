{
  "name": "@ponticus/escomplex-commons",
  "version": "0.0.1",
  "homepage": "https://github.com/gabrielcsapo/@ponticus/escomplex-commons/",
  "description": "Provides core common utilities for @ponticus/escomplex modules and plugins.",
  "main": "./dist/index.js",
  "types": "./dist/dts/index.d.ts",
  "scripts": {
    "build": "swc src -d dist --config-file ../../.swcrc",
    "build:types": "tsc",
    "lint": "eslint .",
    "prepublish": "npm run build",
    "test": "vitest"
  },
  "keywords": [
    "escomplex",
    "commons",
    "report",
    "traits",
    "formats"
  ],
  "devDependencies": {
    "@ponticus/escomplex-test-data": "0.0.1",
    "@ponticus/plugin-manager": "0.0.1",
    "@swc/cli": "^0.1.57",
    "@swc/core": "^1.3.14",
    "@typescript-eslint/eslint-plugin": "^5.42.1",
    "@typescript-eslint/parser": "^5.42.1",
    "eslint": "^8.27.0",
    "eslint-plugin-prettier": "^4.2.1",
    "eslint-plugin-tsdoc": "^0.2.17",
    "fs-extra": "^10.1.0",
    "vitest": "^0.25.1"
  },
  "volta": {
    "node": "18.12.1"
  }
}
