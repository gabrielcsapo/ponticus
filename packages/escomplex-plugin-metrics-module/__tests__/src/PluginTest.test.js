import fs from "fs";
import path from "path";

import { test, describe, expect } from "vitest";

import { PluginSyntaxBabylon } from "@ponticus/escomplex-plugin-syntax-babylon";
import { ModuleReport, ModuleScopeControl } from "@ponticus/escomplex-commons";

import { ASTWalker } from "@ponticus/ast-walker";

import PluginMetricsModule from "../../dist/PluginMetricsModule";

const pluginData = [{ name: "ESM", PluginClass: PluginMetricsModule }];

pluginData.forEach((plugin) => {
  describe(`(${plugin.name}) plugin:`, () => {
    describe("initialize:", () => {
      const instance = new plugin.PluginClass();

      test("plugin was object", () => {
        expect(typeof instance).toBe("object");
      });

      test("plugin function onConfigure is exported", () => {
        expect(typeof instance.onConfigure).toBe("function");
      });

      test("plugin function onEnterNode is exported", () => {
        expect(typeof instance.onEnterNode).toBe("function");
      });

      test("plugin function onModuleAverage is exported", () => {
        expect(typeof instance.onModuleAverage).toBe("function");
      });

      test("plugin function onModuleCalculate is exported", () => {
        expect(typeof instance.onModuleCalculate).toBe("function");
      });

      test("plugin function onModuleCalculate is exported", () => {
        expect(typeof instance.onModulePostAverage).toBe("function");
      });

      test("plugin function onModulePreScopeCreated is exported", () => {
        expect(typeof instance.onModulePreScopeCreated).toBe("function");
      });

      test("plugin function onModulePostScopeCreated is exported", () => {
        expect(typeof instance.onModulePostScopeCreated).toBe("function");
      });
    });

    describe("method invocation:", () => {
      const instance = new plugin.PluginClass();

      test("plugin throws on empty event data", () => {
        expect(() => {
          instance.onConfigure();
        }).toThrow();
      });

      test("plugin does not throw on proper event data", () => {
        expect(() => {
          instance.onConfigure({ data: { options: {}, settings: {} } });
        }).not.toThrow();
      });

      test("plugin passes back syntax data", () => {
        const event = { data: { options: {}, settings: {} } };
        instance.onConfigure(event);
        expect(event.data.settings.newmi).toBe(false);
      });
    });

    describe("module results:", () => {
      const syntaxInstance = new PluginSyntaxBabylon();
      const instance = new plugin.PluginClass();

      const ast = JSON.parse(
        fs.readFileSync(
          path.resolve(__dirname, "../fixture/espree-estree.json"),
          "utf8"
        )
      );

      /**
       * Bootstraps the ESComplexModule runtime and fudges traversing the AST with the Babylon trait syntaxes.
       *
       * Note: That the control flow below exactly replicates typhonjs-escomplex-module / ESComplexModule. If there
       * are any future changes to ESComplexModule the below control flow will need to be modified accordingly.
       */
      test("verify calculated results", () => {
        const moduleReport = new ModuleReport(
          ast.loc.start.line,
          ast.loc.end.line
        );
        const scopeControl = new ModuleScopeControl(moduleReport);

        let event = { data: { options: {}, settings: {} } };

        instance.onConfigure(event);
        syntaxInstance.onConfigure(event);

        const settings = event.data.settings;

        event = { data: { settings, syntaxes: {} } };

        syntaxInstance.onLoadSyntax(event);

        const syntaxes = event.data.syntaxes;

        // Completely traverse the provided AST and defer to plugins to process node traversal.
        new ASTWalker().traverse(ast, {
          enterNode: (node, parent) => {
            const syntax = syntaxes[node.type];

            let ignoreKeys = [];

            // Process node syntax / ignore keys.
            if (typeof syntax === "object") {
              if (syntax.ignoreKeys) {
                ignoreKeys = syntax.ignoreKeys.valueOf(node, parent);
              }
            }

            ignoreKeys = instance.onEnterNode({
              data: {
                moduleReport,
                scopeControl,
                ignoreKeys,
                syntaxes,
                settings,
                node,
                parent,
              },
            });

            // Process node syntax / create scope.
            if (typeof syntax === "object") {
              if (syntax.ignoreKeys) {
                ignoreKeys = syntax.ignoreKeys.valueOf(node, parent);
              }

              if (syntax.newScope) {
                const newScope = syntax.newScope.valueOf(node, parent);

                if (newScope) {
                  // TODO: new code
                  instance.onModulePreScopeCreated({
                    data: {
                      moduleReport,
                      scopeControl,
                      newScope,
                      node,
                      parent,
                    },
                  });
                  scopeControl.createScope(newScope);
                  instance.onModulePostScopeCreated({
                    data: { moduleReport, scopeControl, newScope },
                  });

                  // TODO: old code
                  // scopeControl.createScope(newScope);
                  // instance.onModuleScopeCreated({ data: { moduleReport, scopeControl, newScope } });
                }
              }
            }

            return ignoreKeys;
          },

          exitNode: (node, parent) => {
            const syntax = syntaxes[node.type];

            // Process node syntax / pop scope.
            if (typeof syntax === "object" && syntax.newScope) {
              const newScope = syntax.newScope.valueOf(node, parent);

              if (newScope) {
                scopeControl.popScope(newScope);
              }
            }
          },
        });

        instance.onModuleCalculate({
          data: { moduleReport, syntaxes, settings },
        });
        instance.onModuleAverage({
          data: { moduleReport, syntaxes, settings },
        });
        instance.onModulePostAverage({
          data: { moduleReport, syntaxes, settings },
        });

        moduleReport.finalize();

        expect(JSON.stringify(moduleReport)).toMatchSnapshot();
      });
    });
  });
});
