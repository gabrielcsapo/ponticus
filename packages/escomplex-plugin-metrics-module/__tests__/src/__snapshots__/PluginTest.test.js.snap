// Vitest Snapshot v1

exports[`(ESM) plugin: > module results: > verify calculated results 1`] = `
ModuleReport {
  "aggregate": AggregateReport {
    "aggregate": undefined,
    "cyclomatic": 19,
    "cyclomaticDensity": 26.389,
    "halstead": HalsteadData {
      "bugs": 0.66,
      "difficulty": 56.144,
      "effort": 111199.295,
      "length": 317,
      "operands": {
        "distinct": 45,
        "identifiers": [
          "polyBtoa",
          "array",
          "\\"string\\"",
          "str",
          "s_BTOA_NEW_BINARY",
          "length",
          "j",
          "0",
          "ch",
          "charCodeAt",
          "255",
          "Error",
          "\\"Not ascii. Base64.encode can only take ascii strings\\"",
          "Array",
          "isArray",
          "TypeError",
          "polyBtoa error: 'array' is not an 'Array'.",
          "answer",
          "a",
          "null",
          "b",
          "c",
          "d",
          "i",
          "3",
          "2",
          "63",
          "4",
          "1",
          "15",
          "6",
          "push",
          "s_BTOA_GET_CHAR",
          "\\"=\\"",
          "join",
          "\\"\\"",
          "btoa",
          "self",
          "\\"object\\"",
          "s_BTOA_BASE_64_CHARS",
          "\\"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/\\"",
          "val",
          "charAt",
          "len",
          "ret",
        ],
        "total": 163,
      },
      "operators": {
        "distinct": 31,
        "identifiers": [
          "const",
          "=",
          "function=>",
          "if",
          "===",
          "typeof (prefix)",
          "()",
          ".",
          "for",
          "let",
          "<",
          "++ (postfix)",
          ">",
          "throw",
          "new",
          "[]",
          "! (prefix)",
          "\`\`",
          "switch",
          "%",
          "case",
          "&",
          ">>",
          "<<",
          "break",
          "|=",
          "!==",
          "else",
          "return",
          ":?",
          "&&",
        ],
        "total": 154,
      },
      "time": 6177.739,
      "vocabulary": 76,
      "volume": 1980.593,
    },
    "paramCount": 3,
    "sloc": {
      "logical": 72,
      "physical": 102,
    },
  },
  "aggregateAverage": MethodAverage {
    "cyclomatic": 4.75,
    "cyclomaticDensity": 6.597,
    "halstead": HalsteadAverage {
      "bugs": 0.165,
      "difficulty": 14.036,
      "effort": 27799.824,
      "length": 79.25,
      "operands": {
        "distinct": 11.25,
        "total": 40.75,
      },
      "operators": {
        "distinct": 7.75,
        "total": 38.5,
      },
      "time": 1544.435,
      "vocabulary": 19,
      "volume": 495.148,
    },
    "paramCount": 0.75,
    "sloc": {
      "logical": 18,
      "physical": 25.5,
    },
  },
  "classes": [],
  "dependencies": [],
  "errors": [],
  "filePath": undefined,
  "lineEnd": 112,
  "lineStart": 11,
  "maintainability": 87.622,
  "methodAverage": MethodAverage {
    "cyclomatic": 5,
    "cyclomaticDensity": 40.23,
    "halstead": HalsteadAverage {
      "bugs": 0.176,
      "difficulty": 21.167,
      "effort": 25975.571,
      "length": 91.667,
      "operands": {
        "distinct": 14.333,
        "total": 47.667,
      },
      "operators": {
        "distinct": 13.667,
        "total": 44,
      },
      "time": 1443.087,
      "vocabulary": 28,
      "volume": 526.64,
    },
    "paramCount": 1,
    "sloc": {
      "logical": 21.333,
      "physical": 29,
    },
  },
  "methods": [
    ModuleMethodReport {
      "aggregate": undefined,
      "cyclomatic": 12,
      "cyclomaticDensity": 20.69,
      "errors": [],
      "halstead": HalsteadData {
        "bugs": 0.494,
        "difficulty": 52,
        "effort": 77083.002,
        "length": 248,
        "operands": {
          "distinct": 35,
          "identifiers": [
            "array",
            "\\"string\\"",
            "str",
            "s_BTOA_NEW_BINARY",
            "length",
            "j",
            "0",
            "ch",
            "charCodeAt",
            "255",
            "Error",
            "\\"Not ascii. Base64.encode can only take ascii strings\\"",
            "Array",
            "isArray",
            "TypeError",
            "polyBtoa error: 'array' is not an 'Array'.",
            "answer",
            "a",
            "null",
            "b",
            "c",
            "d",
            "i",
            "3",
            "2",
            "63",
            "4",
            "1",
            "15",
            "6",
            "push",
            "s_BTOA_GET_CHAR",
            "\\"=\\"",
            "join",
            "\\"\\"",
          ],
          "total": 130,
        },
        "operators": {
          "distinct": 28,
          "identifiers": [
            "if",
            "===",
            "typeof (prefix)",
            "const",
            "=",
            "()",
            ".",
            "for",
            "let",
            "<",
            "++ (postfix)",
            ">",
            "throw",
            "new",
            "[]",
            "! (prefix)",
            "\`\`",
            "switch",
            "%",
            "case",
            "&",
            ">>",
            "<<",
            "break",
            "|=",
            "!==",
            "else",
            "return",
          ],
          "total": 118,
        },
        "time": 4282.389,
        "vocabulary": 63,
        "volume": 1482.365,
      },
      "lineEnd": 87,
      "lineStart": 11,
      "maxNestedMethodDepth": 0,
      "name": "<anon method-1>",
      "nestedMethods": [],
      "paramCount": 1,
      "paramNames": [
        "array",
      ],
      "sloc": {
        "logical": 58,
        "physical": 77,
      },
    },
    ModuleMethodReport {
      "aggregate": undefined,
      "cyclomatic": 1,
      "cyclomaticDensity": 50,
      "errors": [],
      "halstead": HalsteadData {
        "bugs": 0.005,
        "difficulty": 1.5,
        "effort": 23.265,
        "length": 6,
        "operands": {
          "distinct": 3,
          "identifiers": [
            "s_BTOA_BASE_64_CHARS",
            "charAt",
            "val",
          ],
          "total": 3,
        },
        "operators": {
          "distinct": 3,
          "identifiers": [
            "return",
            "()",
            ".",
          ],
          "total": 3,
        },
        "time": 1.292,
        "vocabulary": 6,
        "volume": 15.51,
      },
      "lineEnd": 105,
      "lineStart": 102,
      "maxNestedMethodDepth": 0,
      "name": "<anon method-2>",
      "nestedMethods": [],
      "paramCount": 1,
      "paramNames": [
        "val",
      ],
      "sloc": {
        "logical": 2,
        "physical": 4,
      },
    },
    ModuleMethodReport {
      "aggregate": undefined,
      "cyclomatic": 2,
      "cyclomaticDensity": 50,
      "errors": [],
      "halstead": HalsteadData {
        "bugs": 0.027,
        "difficulty": 10,
        "effort": 820.447,
        "length": 21,
        "operands": {
          "distinct": 5,
          "identifiers": [
            "ret",
            "i",
            "0",
            "len",
            "push",
          ],
          "total": 10,
        },
        "operators": {
          "distinct": 10,
          "identifiers": [
            "const",
            "=",
            "[]",
            "for",
            "let",
            "<",
            "++ (postfix)",
            "()",
            ".",
            "return",
          ],
          "total": 11,
        },
        "time": 45.58,
        "vocabulary": 15,
        "volume": 82.045,
      },
      "lineEnd": 112,
      "lineStart": 107,
      "maxNestedMethodDepth": 0,
      "name": "<anon method-3>",
      "nestedMethods": [],
      "paramCount": 1,
      "paramNames": [
        "len",
      ],
      "sloc": {
        "logical": 4,
        "physical": 6,
      },
    },
  ],
  "settings": {},
  "srcPath": undefined,
  "srcPathAlias": undefined,
}
`;
